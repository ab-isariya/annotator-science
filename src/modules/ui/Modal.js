import {useCallback, useEffect, useRef} from 'react';
import PropTypes from 'prop-types';

import {blackToGrey500, toBlack} from '@utils/colorsFilters';
import {ReactComponent as Close} from '@assets/svgs/Close.svg';

const Modal = ({open, toggleModal, children}) => {
  const wrapperRef = useRef(null);

  /**
   * Execute onClose function on pressing Esc
   *
   * @param event
   */
  const handlerEsc = useCallback(
    (event) => {
      // Esc
      if (event.keyCode === 27) {
        toggleModal(false);
      }
    },
    [toggleModal]
  );

  /**
   *
   */
  useEffect(() => {
    document.addEventListener('keydown', handlerEsc, false);

    return () => {
      document.removeEventListener('keydown', handlerEsc, false);
    };
  }, [handlerEsc]);

  return (
    // Modal background
    open && (
      <div
        data-testid="modal"
        className="fixed z-10 inset-0 overflow-y-auto flex">
        <div
          className="relative p-8 bg-white w-full max-w-lg m-auto flex-col flex rounded-md border border-grey-500"
          ref={wrapperRef}>
          <span className="absolute top-0 right-0 p-4">
            <Close
              onClick={() => toggleModal(false)}
              style={{filter: `${toBlack} ${blackToGrey500}`}}
            />
          </span>
          <div>{children}</div>
        </div>
      </div>
    )
  );
};

Modal.propTypes = {
  //Children to render inside of modal
  children: PropTypes.node,
  //If the modal is open or not
  open: PropTypes.bool.isRequired,
  //Toggle for opening/closing the modal.
  toggleModal: PropTypes.func.isRequired
};

export default Modal;
