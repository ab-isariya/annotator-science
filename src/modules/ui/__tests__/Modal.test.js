import Modal from '../Modal';

import renderer from 'react-test-renderer';
import {render} from '@testing-library/react';
import {fireEvent} from '@testing-library/dom';
import {getByTestId} from '@testing-library/dom';
import {shallow} from 'enzyme';

describe('<Modal/>', () => {
  it('Renders without crashing', () => {
    const comp = renderer
      .create(<Modal toggleModal={jest.fn()} open={false} />)
      .toJSON();

    expect(comp).toMatchSnapshot();
  });

  it('Renders nothing when open is false', () => {
    const comp = renderer
      .create(<Modal toggleModal={jest.fn()} open={false} />)
      .toJSON();

    expect(comp).toMatchSnapshot();
    expect(comp).toBe(null);
  });

  it('Renders the modal when open is true', () => {
    const comp = renderer
      .create(<Modal toggleModal={jest.fn()} open={true} />)
      .toJSON();

    expect(comp).toMatchSnapshot();
    expect(comp).not.toBe(null);
  });

  //TODO(Rejon): Come back to this, there's a problem with document event listeners not going through for fake keypresses.
  //			   Specifically, the wrapper we past to spyOn is either not getting the event listener or it receives 'undefined'.
  it('Handles the escape key correctly', () => {
    // let events = {};
    // document.addEventListener = jest.fn((event, cb) => {
    // 	events[event] = cb;
    // });
    // const doSomething = jest.fn();
    // const { container } = render(<Modal toggleModal={doSomething} open={true} />);
    // const wrapper = getByTestId(container, 'modal');
    // const spy = jest.spyOn(wrapper, 'handlerEsc');
    // events.keydown({ keycode: 27 });
    // expect(spy).toHaveBeenCalled();
  });
});
