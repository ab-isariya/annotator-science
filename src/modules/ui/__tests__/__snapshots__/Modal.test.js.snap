// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Modal/> Renders nothing when open is false 1`] = `null`;

exports[`<Modal/> Renders the modal when open is true 1`] = `
<div
  className="fixed z-10 inset-0 overflow-y-auto flex"
  data-testid="modal"
>
  <div
    className="relative p-8 bg-white w-full max-w-lg m-auto flex-col flex rounded-md border border-grey-500"
  >
    <span
      className="absolute top-0 right-0 p-4"
    >
      <svg
        onClick={[Function]}
        style={
          Object {
            "filter": "brightness(0) saturate(100%) invert(42%) sepia(23%) saturate(354%) hue-rotate(182deg) brightness(94%) contrast(85%)",
          }
        }
      >
        Close.svg
      </svg>
    </span>
    <div />
  </div>
</div>
`;

exports[`<Modal/> Renders without crashing 1`] = `null`;
